File: java/javasearch/src/test/java/javasearch/SearchResultTest.java
package javasearch;

import org.junit.Test;

import java.util.Arrays;
import java.util.List;
import java.util.regex.Pattern;

import static org.junit.Assert.assertEquals;

public class SearchResultTest {

    public SearchResultTest() {

    }

    @Test
    public final void testSingleLineSearchResult() {
        final Pattern pattern = Pattern.compile("Search");
        final SearchFile searchFile = new SearchFile("~/src/git/xsearch/csharp/CsSearch/CsSearch",
            "Searcher.cs", FileType.TEXT);
        final int lineNum = 10;
        final int matchStartIndex = 15;
        final int matchEndIndex = 23;
        final String line = "\tpublic class Searcher\n";
        final SearchResult searchResult = new SearchResult(pattern, searchFile, lineNum,
            matchStartIndex, matchEndIndex, line);
        final String expectedPath = "~/src/git/xsearch/csharp/CsSearch/CsSearch/Searcher.cs";
        final String expectedOutput = String.format("%s: %d: [%d:%d]: %s", expectedPath,
                lineNum, matchStartIndex, matchEndIndex, line.trim());
        assertEquals(searchResult.toString(), expectedOutput);
    }

    @Test
    public final void testBinaryFileSearchResult() {
        final Pattern pattern = Pattern.compile("Search");
        final SearchFile searchFile = new SearchFile("~/src/git/xsearch/csharp/CsSearch/CsSearch",
            "Searcher.exe", FileType.BINARY);
        final int lineNum = 0;
        final int matchStartIndex = 0;
        final int matchEndIndex = 0;
        final SearchResult searchResult = new SearchResult(pattern, searchFile, lineNum,
            matchStartIndex, matchEndIndex, null);
        final String expectedPath = "~/src/git/xsearch/csharp/CsSearch/CsSearch/Searcher.exe";
        final String expectedOutput = String.format("%s matches", expectedPath);
        assertEquals(searchResult.toString(), expectedOutput);
    }

    @Test
    public final void testMultiLineSearchResult() {
        final Pattern pattern = Pattern.compile("Search");
        final SearchFile searchFile = new SearchFile("~/src/git/xsearch/csharp/CsSearch/CsSearch",
                "Searcher.cs", FileType.TEXT);
        final int lineNum = 10;
        final int matchStartIndex = 15;
        final int matchEndIndex = 23;
        final String line = "\tpublic class Searcher\n";
        final List<String> linesBefore = Arrays.asList("namespace CsSearch\n", "{\n");
        final List<String> linesAfter = Arrays.asList("\t{\n", "\t\tprivate readonly FileTypes _fileTypes;\n");
        final SearchResult searchResult = new SearchResult(pattern, searchFile, lineNum,
                matchStartIndex, matchEndIndex, line, linesBefore, linesAfter);
        final String expectedPath = "~/src/git/xsearch/csharp/CsSearch/CsSearch/Searcher.cs";
        final String expectedOutput = String.format(
                "================================================================================\n" +
                "%s: %d: [%d:%d]\n" +
                "--------------------------------------------------------------------------------\n" +
                "   8 | namespace CsSearch\n" +
                "   9 | {\n" +
                "> 10 | \tpublic class Searcher\n" +
                "  11 | \t{\n" +
                "  12 | \t\tprivate readonly FileTypes _fileTypes;\n",
                expectedPath, lineNum, matchStartIndex, matchEndIndex);
        assertEquals(searchResult.toString(), expectedOutput);
    }
}

File: kotlin/ktsearch/src/test/kotlin/ktsearch/SearchResultTest.kt
package ktsearch

import org.junit.Assert.assertEquals
import org.junit.Test
import java.io.File
import java.util.*

/**
 * @author cary on 7/30/16.
 */
class SearchResultTest {

    @Test
    fun testSingleLineSearchResult() {
        val settings = getDefaultSettings().copy(colorize = false)
        val formatter = SearchResultFormatter(settings)
        val pattern = Regex("Search")
        val file = File("~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.cs")
        val searchFile = SearchFile(file, FileType.CODE)
        val lineNum = 10
        val matchStartIndex = 15
        val matchEndIndex = 23
        val line = "\tpublic class Searcher\n"
        val searchResult = SearchResult(pattern, searchFile, lineNum,
                matchStartIndex, matchEndIndex, line)
        val expectedPath = "~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.cs"
        val expectedOutput = String.format("%s: %d: [%d:%d]: %s", expectedPath,
                lineNum, matchStartIndex, matchEndIndex, line.trim { it <= ' ' })
        val output = formatter.format(searchResult)
        assertEquals(expectedOutput, output)
    }

    @Test
    fun testSingleLineLongerThanMaxLineLengthSearchResult() {
        val settings = getDefaultSettings().copy(colorize = false, maxLineLength = 100)
        val formatter = SearchResultFormatter(settings)
        val pattern = Regex("maxlen")
        val file = File("./maxlen.txt")
        val searchFile = SearchFile(file, FileType.TEXT)
        val lineNum = 1
        val matchStartIndex = 53
        val matchEndIndex = 59
        val line = "0123456789012345678901234567890123456789012345678901maxlen8901234567890123456789012345678901234567890123456789"
        val linesBeforeAfter: List<String> = ArrayList()
        val searchResult = SearchResult(pattern, searchFile, lineNum, matchStartIndex, matchEndIndex,
                line, linesBeforeAfter, linesBeforeAfter)
        val expectedPath = "." + File.separator + "maxlen.txt"
        val expectedLine = "...89012345678901234567890123456789012345678901maxlen89012345678901234567890123456789012345678901..."
        val expectedOutput = String.format("%s: %d: [%d:%d]: %s", expectedPath,
                lineNum, matchStartIndex, matchEndIndex, expectedLine)
        val output = formatter.format(searchResult)
        assertEquals(expectedOutput, output)
    }

    @Test
    fun testSingleLineLongerColorizeSearchResult() {
        val settings = getDefaultSettings().copy(colorize = true, maxLineLength = 100)
        val formatter = SearchResultFormatter(settings)
        val pattern = Regex("maxlen")
        val file = File("./maxlen.txt")
        val searchFile = SearchFile(file, FileType.TEXT)
        val lineNum = 1
        val matchStartIndex = 53
        val matchEndIndex = 59
        val line = "0123456789012345678901234567890123456789012345678901maxlen8901234567890123456789012345678901234567890123456789"
        val linesBeforeAfter: List<String> = ArrayList()
        val searchResult = SearchResult(pattern, searchFile, lineNum, matchStartIndex, matchEndIndex,
                line, linesBeforeAfter, linesBeforeAfter)
        val expectedPath = "." + File.separator + "maxlen.txt"
        val expectedLine = "...89012345678901234567890123456789012345678901" +
                Color.GREEN +
                "maxlen" +
                Color.RESET +
                "89012345678901234567890123456789012345678901..."
        val expectedOutput = String.format("%s: %d: [%d:%d]: %s", expectedPath,
                lineNum, matchStartIndex, matchEndIndex, expectedLine)
        val output = formatter.format(searchResult)
        assertEquals(expectedOutput, output)
    }

    @Test
    fun testBinaryFileSearchResult() {
        val settings = getDefaultSettings()
        val formatter = SearchResultFormatter(settings)
        val pattern = Regex("Search")
        val file = File("~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.exe")
        val searchFile = SearchFile(file, FileType.BINARY)
        val lineNum = 0
        val matchStartIndex = 0
        val matchEndIndex = 0
        val searchResult = SearchResult(pattern, searchFile, lineNum,
                matchStartIndex, matchEndIndex, "")
        val expectedPath = "~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.exe"
        val expectedOutput = String.format("%s matches at [0:0]", expectedPath)
        val output = formatter.format(searchResult)
        assertEquals(expectedOutput, output)
    }

    @Test
    fun testMultiLineSearchResult() {
        val settings = getDefaultSettings().copy(colorize = false)
        val formatter = SearchResultFormatter(settings)
        val pattern = Regex("Search")
        val file = File("~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.cs")
        val searchFile = SearchFile(file, FileType.TEXT)
        val lineNum = 10
        val matchStartIndex = 15
        val matchEndIndex = 23
        val line = "\tpublic class Searcher"
        val linesBefore = listOf("namespace CsSearch", "{")
        val linesAfter = listOf("\t{", "\t\tprivate readonly FileTypes _fileTypes;")
        val searchResult = SearchResult(pattern, searchFile, lineNum,
                matchStartIndex, matchEndIndex, line, linesBefore, linesAfter)
        val expectedPath = "~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.cs"
        val expectedOutput =
                """================================================================================
                   |$expectedPath: $lineNum: [$matchStartIndex:$matchEndIndex]
                   |--------------------------------------------------------------------------------
                   |   8 | namespace CsSearch
                   |   9 | {
                   |> 10 | 	public class Searcher
                   |  11 | 	{
                   |  12 | 		private readonly FileTypes _fileTypes;
                   |""".trimMargin()
        val output = formatter.format(searchResult)
        assertEquals(expectedOutput, output)
    }
}
